core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java

```diff
@@ -228,110 +232,124 @@ private void doCancel() {
 
     @Override
     SubscriptionImpl subscribe(SubscriptionImpl subscription) {
-        if (!subscriptionUpdater.compareAndSet(this, null, subscription)) {
-            final SubscriptionImpl oldSubscription = this.subscription;
+        if (!downstreamSubscriptionUpdater.compareAndSet(this, null, subscription)) {
+            final SubscriptionImpl oldSubscription = downstreamSubscription;
             assert oldSubscription != null;
             return oldSubscription;
         }
 
         final Subscriber<Object> subscriber = subscription.subscriber();
         if (subscription.needsDirectInvocation()) {
-            subscriber.onSubscribe(subscription);
-            safeOnSubscribeToDelegate();
+            subscribe0(subscription, subscriber);
         } else {
             subscription.executor().execute(() -> {
-                subscriber.onSubscribe(subscription);
-                safeOnSubscribeToDelegate();
+                subscribe0(subscription, subscriber);
             });
         }
 
         return subscription;
     }
 
-    private void safeOnSubscribeToDelegate() {
-        if (delegate == null || subscription == null) {
+    private void subscribe0(SubscriptionImpl subscription, Subscriber<Object> subscriber) {
+        try {
+            subscriber.onSubscribe(subscription);
+        } catch (Exception e) {
+            abort(e);
+            return;
+        }
+        safeOnSubscribeToUpstream();
+    }
+
+    private void safeOnSubscribeToUpstream() {
+        final StreamMessage<T> upstream = this.upstream;
+        final SubscriptionImpl downstreamSubscription = this.downstreamSubscription;
+        if (upstream == null || downstreamSubscription == null) {
             return;
         }
 
-        if (!subscribedToDelegateUpdater.compareAndSet(this, 0, 1)) {
+        if (!subscribedToUpstreamUpdater.compareAndSet(this, 0, 1)) {
             return;
         }
 
         final Builder<SubscriptionOption> builder = ImmutableList.builder();
-        if (subscription.withPooledObjects()) {
+        if (downstreamSubscription.withPooledObjects()) {
             builder.add(SubscriptionOption.WITH_POOLED_OBJECTS);
         }
-        if (subscription.notifyCancellation()) {
+        if (downstreamSubscription.notifyCancellation()) {
             builder.add(SubscriptionOption.NOTIFY_CANCELLATION);
         }
 
-        delegate.subscribe(new ForwardingSubscriber(),
-                           subscription.executor(),
+        upstream.subscribe(new ForwardingSubscriber(downstreamSubscription.subscriber()),
+                           downstreamSubscription.executor(),
```

Comment:
Could you explain why we need to pass them? Can't we access them via `downstreamSubscription`, since `ForwardingSubscriber` is not a static class?